File: advent2017/extras/cxx/src/days/day09.cpp
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "860cd63e00136c29310e25db6f4f1573a2b2574598dc72f44a6308ddf5a967c3";
const std::string kExpectedPart1 = "21037";
const std::string kExpectedPart2 = "9495";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{"advent2017/Day9/d9_input.txt", "Day9/d9_input.txt", "../Day9/d9_input.txt", "../../Day9/d9_input.txt"};
    for (const auto& c : cands) if (std::filesystem::exists(c)) return c;
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

std::string trim_ws(const std::string& s) {
    std::size_t b = 0;
    while (b < s.size() && std::isspace(static_cast<unsigned char>(s[b]))) b++;
    std::size_t e = s.size();
    while (e > b && std::isspace(static_cast<unsigned char>(s[e - 1]))) e--;
    return s.substr(b, e - b);
}

std::pair<long long, long long> scan_stream(const std::string& raw) {
    const std::string s = trim_ws(raw);
    long long depth = 0;
    long long score = 0;
    long long garbage = 0;
    bool in_garbage = false;

    for (std::size_t i = 0; i < s.size(); ++i) {
        const char c = s[i];
        if (in_garbage) {
            if (c == '!') {
                i++;
                continue;
            }
            if (c == '>') {
                in_garbage = false;
            } else {
                garbage++;
            }
            continue;
        }

        if (c == '<') {
            in_garbage = true;
        } else if (c == '{') {
            depth++;
            score += depth;
        } else if (c == '}') {
            depth--;
        }
    }

    return {score, garbage};
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");

    auto t0 = std::chrono::steady_clock::now();
    const auto [p1, p2] = scan_stream(raw);
    const std::string ans = (part == 1) ? std::to_string(p1) : std::to_string(p2);
    const std::string expected = (part == 1) ? kExpectedPart1 : kExpectedPart2;
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=9 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day09.rs
// Architecture Notes:
// - This file is heavily commented for long-term maintainability and reconstruction.
// - Pipeline shape is parse -> model -> compute -> emit.
// - Performance-sensitive sections document data-layout and concurrency tradeoffs.

use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "860cd63e00136c29310e25db6f4f1573a2b2574598dc72f44a6308ddf5a967c3";
const EXPECTED_PART1: &str = "21037";
const EXPECTED_PART2: &str = "9495";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() { return provided.to_string(); }
    let cands = ["advent2017/Day9/d9_input.txt", "Day9/d9_input.txt", "../Day9/d9_input.txt", "../../Day9/d9_input.txt"];
    for c in cands { if Path::new(c).exists() { return c.to_string(); } }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new(); h.update(&b); hex::encode(h.finalize())
}

fn scan_stream(raw: &str) -> (i64, i64) {
    let s = raw.trim().as_bytes();
    let mut depth: i64 = 0;
    let mut score: i64 = 0;
    let mut garbage: i64 = 0;
    let mut in_garbage = false;
    let mut i = 0usize;

    while i < s.len() {
        let c = s[i];
        if in_garbage {
            if c == b'!' {
                i += 2;
                continue;
            }
            if c == b'>' {
                in_garbage = false;
            } else {
                garbage += 1;
            }
            i += 1;
            continue;
        }

        if c == b'<' {
            in_garbage = true;
        } else if c == b'{' {
            depth += 1;
            score += depth;
        } else if c == b'}' {
            depth -= 1;
        }
        i += 1;
    }

    (score, garbage)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required"); }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA { panic!("checksum mismatch"); }
    let raw = fs::read_to_string(&in_path).expect("read_to_string");

    let t0 = Instant::now();
    let (p1, p2) = scan_stream(&raw);
    let ans = if part == 1 { p1.to_string() } else { p2.to_string() };
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected { panic!("answer mismatch"); }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=9 part={} runtime_ms={:.3}", part, ms);
}
